operator/controllers/credential_controller.go

```diff
@@ -0,0 +1,572 @@
+			return ctrl.Result{}, err
+		}
+		if err := r.Update(ctx, credential); err != nil {
+			return ctrl.Result{}, err
+		}
+		// TODO: make this a constant.
+		return ctrl.Result{RequeueAfter: 3 * time.Second}, nil
+	}
+
+	secret, err := r.getOrCreateSecret(ctx, &req, credential)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if secret == nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	plan := &servicebrokerv1alpha1.Plan{}
+	if err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      instance.Spec.Plan,
+			Namespace: req.Namespace,
+		},
+		plan,
+	); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return r.runBindingPod(ctx, &req, credential, instance, plan, secret.Name)
+}
+
+// getOrCreateSecret returns or creates the existing secret referenced by the credential. When the
+// secret and the error returning values are nil, it means the secret was created. A requeue should
+// be issued for the next reconcile loop to get the current secret from the API server.
+func (r *CredentialReconciler) getOrCreateSecret(
+	ctx context.Context,
+	req *ctrl.Request,
+	credential *servicebrokerv1alpha1.Credential,
+) (*corev1.Secret, error) {
+	desired := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      credential.SecretRef.Name,
+			Namespace: req.Namespace,
+			// TODO: add proper labels.
+		},
+		Data: map[string][]byte{},
+	}
+	if err := ctrl.SetControllerReference(credential, desired, r.scheme); err != nil {
+		return nil, err
+	}
+	current := &corev1.Secret{}
+	if err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      credential.SecretRef.Name,
+			Namespace: req.Namespace,
+		},
+		current,
+	); err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		if err := r.Create(ctx, desired); err != nil {
+			return nil, err
+		}
+		return nil, nil
+	}
+	// TODO: if the current secret is not owned by the credential, fail.
+	return current, nil
+}
+
+func (r *CredentialReconciler) runBindingPod(
+	ctx context.Context,
+	req *ctrl.Request,
+	credential *servicebrokerv1alpha1.Credential,
+	instance *servicebrokerv1alpha1.Instance,
+	plan *servicebrokerv1alpha1.Plan,
+	credentialsSecretName string,
+) (ctrl.Result, error) {
+	desiredServiceAccount := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+			// TODO: add proper labels.
+		},
+	}
+	if err := ctrl.SetControllerReference(credential, desiredServiceAccount, r.scheme); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+	}
+
+	currentServiceAccount := &corev1.ServiceAccount{}
+	if err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      desiredServiceAccount.Name,
+			Namespace: desiredServiceAccount.Namespace,
+		},
+		currentServiceAccount,
+	); err != nil {
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to bind: %w", err)
+		}
+		if err := r.Create(ctx, desiredServiceAccount); err != nil {
```

Comment:
Do we ever set the controller reference here?

Reply:
Yes, line 184.